File: haskell/lib/Categorical_combinators.hs
data Term a = Var Int | App (Term a) (Term a) | Lam String (Term a)

eval :: Term a -> [(String, a)] -> a
eval (Var n) env = env !! n
eval (App t1 t2) env = let v1 = eval t1 env in let v2 = eval t2 env in v1 v2
eval (Lam x t) env = \v -> eval t ((x, v) : env)

fix :: (Term a -> Term a) -> Term a
fix f = Lam "x" (App (f (Var 0)) (Var 0))

id = Lam "x" (Var 0)

k = Lam "x" (Lam "y" (Var 0))

b = Lam "x" (Lam "y" (Lam "z" (Var 1)))

File: go/categorical_combinators.go
package efficient

import (
	"fmt"
)

type Term interface {
	Eval(env map[string]interface{}) interface{}
}

type Var struct {
	Index int
}

func (v Var) Eval(env map[string]interface{}) interface{} {
	return env[fmt.Sprintf("%d", v.Index)]
}

type App struct {
	Fn Term
	Arg Term
}

func (a App) Eval(env map[string]interface{}) interface{} {
	fn := a.Fn.Eval(env).(func(interface{}) interface{})
	return fn(a.Arg.Eval(env))
}

type Lam struct {
	Var  string
	Body Term
}

func (l Lam) Eval(env map[string]interface{}) interface{} {
	return func(arg interface{}) interface{} {
		newEnv := make(map[string]interface{})
		for k, v := range env {
			newEnv[k] = v
		}
		newEnv[l.Var] = arg
		return l.Body.Eval(newEnv)
	}
}

func fix(t func(Term) Term) Term {
	return Lam{"x", App(t(Var{0}), Var{0})}
}

func id() func(interface{}) interface{} {
	return func(x interface{}) interface{} { return x }
}

func k() func(interface{}) func(interface{}) interface{} {
	return func(x interface{}) func(interface{}) interface{} {
		return func(y interface{}) interface{} { return x }
	}
}

func b() func(interface{}) func(interface{}) func(interface{}) interface{} {
	return func(x interface{}) func(interface{}) func(interface{}) interface{} {
		return func(y interface{}) func(interface{}) interface{} {
			return func(z interface{}) interface{} { return y }
		}
	}
}